text/0000-proc-macros.md

```diff
@@ -0,0 +1,452 @@
+[design by contract](https://github.com/nrc/libhoare). They allow the ultimate
+flexibility in syntactic abstraction, and offer possibilities for efficiently
+using Rust in novel ways.
+
+Procedural macros are currently unstable and are awkward to define. We would
+like to remedy this by implementing a new, simpler system for procedural macros,
+and for this new system to be on the usual path to stabilisation.
+
+One major problem with the current system is that since it is based on ASTs, if
+we change the Rust language (even in a backwards compatible way) we can easily
+break procedural macros. Therefore, offering the usual backwards compatibility
+guarantees to procedural macros, would inhibit our ability to evolve the
+language. By switching to a token-based (rather than AST- based) system, we hope
+to avoid this problem.
+
+# Detailed design
+[design]: #detailed-design
+
+There are two kinds of procedural macro: function-like and attribute-like. These
+two kinds exist today, and other than naming (see
+[RFC 1561](https://github.com/rust-lang/rfcs/pull/1561)) the syntax for using
+these macros remains unchanged. If the macro is called `foo`, then a function-
+like macro is used with syntax `foo!(...)`, and an attribute-like macro with
+`#[foo(...)] ...`. Macros may be used in the same places as `macro_rules` macros
+and this remains unchanged.
+
+To define a procedural macro, the programmer must write a function with a
+specific signature and attribute. Where `foo` is the name of a function-like
+macro:
+
+```
+#[macro]
+pub fn foo(TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+The first argument is the tokens between the delimiters in the macro use.
+For example in `foo!(a, b, c)`, the first argument would be `[Ident(a), Comma,
+Ident(b), Comma, Ident(c)]`.
+
+The value returned replaces the macro use.
+
+Attribute-like:
+
+```
+#[macro_attribute]
+pub fn foo(Option<TokenStream>, TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+The first argument is a list of the tokens between the delimiters in the macro
+use. Examples:
+
+* `#[foo]` => `None`
+* `#[foo()]` => `Some([])`
+* `#[foo(a, b, c)]` => `Some([Ident(a), Comma, Ident(b), Comma, Ident(c)])`
+
+The second argument is the tokens for the AST node the attribute is placed on.
+Note that in order to compute the tokens to pass here, the compiler must be able
+to parse the code the attribute is applied to. However, the AST for the node
+passed to the macro is discarded, it is not passed to the macro nor used by the
+compiler (in practice, this might not be 100% true due to optimisiations). If
+the macro wants an AST, it must parse the tokens itself.
+
+The attribute and the AST node it is applied to are both replaced by the
+returned tokens. In most cases, the tokens returned by a procedural macro will
+be parsed by the compiler. It is the procedural macro's responsibility to ensure
+that the tokens parse without error. In some cases, the tokens will be consumed
```

Comment:
What happens if the tokens don't parse without error? Does the user of a macro just get a (possibly cryptic) error message, or you envision something more advanced?